Sources/GRPC/Compression/Zlib.swift

```diff
@@ -0,0 +1,422 @@
+      self.zstream.avail_out = 0
+
+      self.zstream.zalloc = nil
+      self.zstream.zfree = nil
+      self.zstream.opaque = nil
+    }
+
+    var availableInputBytes: Int {
+      get {
+        return Int(self.zstream.avail_in)
+      }
+      set {
+        self.zstream.avail_in = UInt32(newValue)
+      }
+    }
+
+    var nextInputBuffer: UnsafeMutablePointer<Bytef>! {
+      get {
+        return self.zstream.next_in
+      }
+      set {
+        self.zstream.next_in = newValue
+      }
+    }
+
+    var availableOutputBytes: Int {
+      get {
+        return Int(self.zstream.avail_out)
+      }
+      set {
+        return self.zstream.avail_out = UInt32(newValue)
+      }
+    }
+
+    var nextOutputBuffer: UnsafeMutablePointer<Bytef>! {
+      get {
+        return self.zstream.next_out
+      }
+      set {
+        self.zstream.next_out = newValue
+      }
+    }
+
+    var lastErrorMessage: String? {
+      guard let bytes = self.zstream.msg else {
+        return nil
+      }
+      return String(cString: bytes)
+    }
+
+    enum InflateOutcome {
+      /// The data was successfully inflated.
+      case complete
+
+      /// A larger output buffer is required.
+      case outputBufferTooSmall
+    }
+
+    struct InflateResult {
+      var bytesWritten: Int
+      var outcome: InflateOutcome
+    }
+
+    mutating func inflate(
+      outputBuffer: UnsafeMutablePointer<UInt8>,
+      outputBufferSize: Int
+    ) throws -> InflateResult {
+      self.nextOutputBuffer = outputBuffer
+      self.availableOutputBytes = outputBufferSize
+
+      defer {
+        self.nextOutputBuffer = nil
+        self.availableOutputBytes = 0
+      }
+
+      let rc = CGRPCZlib_inflate(&self.zstream, Z_FINISH)
+      let outcome: InflateOutcome
+
+      // Possible return codes:
+      // - Z_OK: some progress has been made
+      // - Z_STREAM_END: the end of the compressed data has been reached and all uncompressed output
+      //   has been produced
+      // - Z_NEED_DICT: a preset dictionary is needed at this point
+      // - Z_DATA_ERROR: the input data was corrupted
+      // - Z_STREAM_ERROR: the stream structure was inconsistent
+      // - Z_MEM_ERROR there was not enough memory
+      // - Z_BUF_ERROR if no progress was possible or if there was not enough room in the output
+      //   buffer when Z_FINISH is used.
+      //
+      // Note that Z_OK is not okay here since we always flush with Z_FINISH and therefore
+      // use Z_STREAM_END as our success criteria.
+
+      switch rc {
+      case Z_STREAM_END:
+        outcome = .complete
+
+      case Z_BUF_ERROR:
+        outcome = .outputBufferTooSmall
+
+      default:
+        throw GRPCError.ZlibCompressionFailure(code: rc, message: self.lastErrorMessage).captureContext()
+      }
+
+      return InflateResult(
+        bytesWritten: outputBufferSize - self.availableOutputBytes,
+        outcome: outcome
+      )
+    }
+
+    mutating func deflate(
```

Comment:
If I understand this correctly, this method is used both for compressing and decompressing streams? Feels a bit weird to me, but I'm not a Zlib expert.

Reply:
Not quite; it's only used for compressing. "deflate" is a bit overloaded though, it refers to the compression format as well as the zlib method to compress something in that format. The `deflate` method can also be used to compress data in the gzip format (which is based on the deflate format).

The same is true for decompression which is done with `inflate`, it can handle "deflate" and "gzip" formatted data.